kernel/util.c

```diff
@@ -46,9 +47,20 @@ void __attribute__ ((noreturn)) qemu_debug_shutdown (void) {
   panic("Shutdown failed!");
 }
 
+static int panic_count = 0;
 void __attribute__ ((noreturn)) panic (const char *s) {
-  puts("\r\nKERNEL PANIC: ");
-  puts(s);
+  panic_count++;
+  if (panic_count == 1) {
+    vga_printf("\r\nKERNEL PANIC: %s\r\n", s);
+    vga_printf("Generating backtrace to COM...\r\n");
+    unwind_from_here();
+  } else if (panic_count == 2) {
+    /* Double panic! No time for anything fancy. */
+    puts("\r\nDOUBLE KERNEL PANIC: ");
+    puts(s);
+  } else {
+    /* Seriously? I give up. */
```

Comment:
This is... not what I was imaging the fix would look like, but whatever.

Can't panic's be arbitrarily nested? Just print out the same message after the first two.